src/libraries/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs

```diff
@@ -1960,6 +1962,51 @@ public static void Xml_TypeWithSpecialCharacterInStringMember()
         Assert.Equal(x.Name, y.Name);
     }
 
+    [Fact]
+#if XMLSERIALIZERGENERATORTESTS
+    // Lack of AssemblyDependencyResolver results in assemblies that are not loaded by path to get
+    // loaded in the default ALC, which causes problems for this test.
+    [SkipOnPlatform(TestPlatforms.Browser, "AssemblyDependencyResolver not supported in wasm")]
+#endif
+    [ActiveIssue("34072", TestRuntimes.Mono)]
+    public static void Xml_TypeInCollectibleALC()
+    {
+        ExecuteAndUnload("SerializableAssembly.dll", "SerializationTypes.SimpleType", out var weakRef);
+
+        for (int i = 0; weakRef.IsAlive && i < 10; i++)
+        {
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+        }
+        Assert.True(!weakRef.IsAlive);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void ExecuteAndUnload(string assemblyfile, string typename, out WeakReference wref)
+    {
+        var fullPath = Path.GetFullPath(assemblyfile);
+        var alc = new TestAssemblyLoadContext("XmlSerializerTests", true, fullPath);
+        wref = new WeakReference(alc);
+
+        // Load assembly by path. By name, and it gets loaded in the default ALC.
+        var asm = alc.LoadFromAssemblyPath(fullPath);
+
+        // Ensure the type loaded in the intended non-Default ALC
+        var type = asm.GetType(typename);
+        Assert.Equal(AssemblyLoadContext.GetLoadContext(type.Assembly), alc);
+        Assert.NotEqual(alc, AssemblyLoadContext.Default);
+
+        // Round-Trip the instance
+        XmlSerializer serializer = new XmlSerializer(type);
+        var obj = Activator.CreateInstance(type);
+        var rtobj = SerializeAndDeserialize(obj, null, () => serializer, true);
+        Assert.NotNull(rtobj);
+        Assert.True(rtobj.Equals(obj));
+
+        alc.Unload();
```

Comment:
Please note that this doesn't actually test that unloading works. In .NET Core this method is basically just a hint to the runtime that it should unload the ALC. The actual unloading is driven by GC. So in order for this to correctly work, it should:
* Run the test in a helper method which is guaranteed not inlined - so that none of its locals are still holding onto references to the ALC (or make sure that everything is assigned null, but compiler might still have some generated locals in theory, so the helper method is more robust).
* Call `alc.Unload`
* Store the alc reference in `WeakReference` and remove all other references to it
* In a loop trigger GC and wait for a bit - and then test if the `WeakReference` disappeared.

Good description of this is for example here: https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability

Reply:
That's actually what we do here. We have to unload in this method while we still have the hard reference to the ALC, and then we check the weak reference once we've left the scope of this method and do our GC loop.

Reply:
You're right - I completely missed that... sorry.